### PR TITLE
Get serial number for astromechs and use dev_node_pattern

### PR DESCRIPTION
**Pull request by JAAlvarado-Montes**
Add code to get the serial number of astromechanics controllers. Also, add the `dev_node_pattern` parameter.

## Description
Get the serial number of astromechanics controllers and add the `dev_node_pattern` parameter so it can be used to search for both astromechanics and birger controllers and initialize them.

## How Has This Been Tested?
The changes were tested locally using real hardware.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/astromechanics.py:23]: This should default to `None` in POCS and then be set via the config.

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/astromechanics.py:39]: The dev node pattern should be `None` in POCS to keep it generic and the values can come from the config.

But, if you are not doing anything with these parameters here and they are already in the super class, you don't really need to specify them, they will just go through with `**kwargs`

```suggestion
                 dev_node_pattern=None,
                 serial_number_pattern=serial_number_pattern,
```

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/astromechanics.py:80]: You are not changing this, but ideally this baudrate should also come from the config.

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/astromechanics.py:93]: I still think the exception should be raised when the `if` fails, not when the serial number check fails.  And can we check it for a valid position response rather than the generic `#`?

File: src/panoptes/pocs/focuser/birger.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/birger.py:66]: Similar to the astromechanics, the `dev_node_pattern` should default to `None` and these parameters probably don't need to be specified.

File: src/panoptes/pocs/focuser/serial.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/serial.py:20]: Default to `None` for unknown things.

```suggestion
    def __init__(self, dev_node_pattern=None, initial_position=None,
                 serial_number_pattern=None, *args, **kwargs):
```

Comment (wtgee) [reply src/panoptes/pocs/focuser/serial.py:20]: For the `serial_number_pattern`, I would create your own `re` object if given a string.

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/astromechanics.py:93]:
```suggestion
                raise error.BadSerialConnection(f"{self.name} not found in {self.port}")
```

File: src/panoptes/pocs/focuser/serial.py
Comment (wtgee) [inline src/panoptes/pocs/focuser/serial.py:66]: This should all be moved out of the class instance and become a class method that is called before object initialization.

See discussion in PR review.

**Review by wtgee — CHANGES_REQUESTED** [review]
I realize this is just a copy/paste job from what was there, but I think if this is not a huge priority time wise then we should take the time to clean up the constructor.

Ideally the `dev_node_pattern` does not happen in an object instance so I'm proposing to make this a class method that would be called prior to object initialization. 

I don't see any advantage to this happening dynamically upon creation, but let me know if there is a valid reason still.  @AnthonyHorton @JAAlvarado-Montes

**Comment by JAAlvarado-Montes** [comment]
Hi @wtgee @AnthonyHorton @danjampro ,

I have addressed all of @wtgee's comments. I'll submit a different PR to `huntsman-config` to add  `baudrate` (birger and astromech); `id_vendor` and `id_product` (only for astromech). The classes are now much more organized and I even find the code more readable.

Everything was tested on Huntsman's lens Tuc with different Astromechanics focusers. Please let me know any suggestion or comment.

PS. Tests keep failing because of the new `pyusb` module that astromechanics.py is using. However, I've added it as a new requirement for POCS as @wtgee indicated, so I'm not sure what is happening.

Cheers,
Jaime A.

File: src/panoptes/pocs/focuser/astromechanics.py
Comment (danjampro) [inline src/panoptes/pocs/focuser/astromechanics.py:25]: Minor detail, but shouldn't these be `_vendor_id` and `_product_id`?

File: src/panoptes/pocs/focuser/birger.py
Comment (danjampro) [inline src/panoptes/pocs/focuser/birger.py:44]: Do we need a `max_retries` for these focusers?

**Review by danjampro — COMMENTED** [review]
Hi @JAAlvarado-Montes looks good. I'm not an expert in the focuser code so I'll leave it to the others to give the final review. The test coverage is showing a bit of a drop, can anything be done about this?

File: src/panoptes/pocs/focuser/birger.py
Comment (JAAlvarado-Montes) [reply src/panoptes/pocs/focuser/birger.py:44]: Yes we do, it is used as the maximum number of times to retry a command when it isn't properly sent

**Comment by lspitler** [comment]
@JAAlvarado-Montes what is the status of this? Do you need a review from @wtgee ?

**Comment by JAAlvarado-Montes** [comment]
@lspitler, as far as I'm concerned all the changes from this PR are ready and working with the Astromechs. I followed all of @wtgee's and @danjampro's comments. Also, tests are passing with no issues.

**Comment by lspitler** [comment]
@wtgee @AnthonyHorton we're about to deploy the AstroMechs in ~7 days. @JAAlvarado-Montes is this PR required for deploying them?

**Review by wtgee — COMMENTED** [review]
@JAAlvarado-Montes I did this review a long time ago but apparently never hit submit, sorry about that!  I think most of the comments still apply but let me know if something doesn't make sense.

**Comment by JAAlvarado-Montes** [comment]
Replaced by #1125

**Pull request closed without merging**
